### PR TITLE
Better integration with backend languages

### PR DESCRIPTION
**Pull request by uros-5**
Fixed completion bug for attributes without values.
```html
<div hx-swap> hello world</p>
```

Improved locality of behavior by adding **`hx-lsp`** attribute. It's easy to understand.

On backend or frontend you simply add comment like this:

```rust
fn hello_world() {
  // hx@example
  println!("hello world");
}

```
Inside templates you call this _'tag'_.

```html
<!-- some nested html -->
<a hx-get="/some_route" hx-lsp="example">hello world</a>
```

Goto definition request for hx-lsp(precise tag name, _'example'_) will redirect client directly to backend function.

We can also specify more tags.
```html
<!-- some nested html -->
<a hx-get="/some_route" hx-lsp="rs-tag js-tag">hello world</a>
```
Of course these tags(comments) must exist on backend/frontend.

I experimented this with JavaScript and Rust. 
Only requirement for tree sitter queries is to have comment nodes inside function node. 
I have prepared queries for Go, Python etc.

Here is quick demo:

https://github.com/ThePrimeagen/htmx-lsp/assets/59397844/99524b5a-389c-4f83-9099-9a26b0e28a0f

If reference array contains only one item location, client will be redirected right there without showing menu(it acts like goto implementation).

Configuration(language server can be used without it, with completion and hover feature):

```json
{ 
  "lang": "rust",
  "template_ext": "jinja",
  "templates": ["./templates"],
  "js_tags": ["./frontend"],
  "backend_tags": ["./backend"]
}

````

### CODE REVIEW

**Comment by cristianoliveira** [comment]
Hello there!

Regarding the fix
> Fixed completion bug for attributes without values.
`<div hx-swap> hello world</p>`

Would you mind cherry picking it to its own PR? I believe it would simplify the review of it.

About the feature proposal, I can see the value in it, and yet, couldn't one achieve the same thing by creating custom tags with https://github.com/universal-ctags/ctags ? Or better its own plugin?

Since those tags aren't part of the HTMX, I'm not sure it belongs to the LSP, it could use a bit more discussion :)

File: src/config.rs [hunk 1]
```diff
@@ -0,0 +1,158 @@
...
+    pub template_ext: String,
+    pub templates: Vec<String>,
+    pub js_tags: Vec<String>,
+    pub backend_tags: Vec<String>,
+}
+
+impl HtmxConfig {
+    pub fn is_backend(&self, ext: &str) -> bool {
+        match self.lang.as_str() {
+            "rust" => ext == "rs",
```
Comment (cristianoliveira) [inline src/config.rs:27]: That's why I believe this feature should be a plugin instead so those who want to support a specific language would create their own implementation

**Comment by MNThomson** [comment]
As @cristianoliveira said, please cherry pick out the fix into another PR. Happy to get that merged!

Honestly, this PR scares me `:)`
- Its effectively an entire rewrite (has enough line changes to be) and is way to big to be able to effectively review (I suck at reviews, so YMMV). Anything over ~150 lines is where my eyes start to glaze over.
- This feels like major feature creep. I really like the idea of having `gf` let me jump to the definition of an endpoint from the frontend, but I'm not sure that this lsp is the place to do so. If there are other solutions that provides tagging and jump points, let use that. Also, having to try and handle issue requests of "please add language x" just isn't manageable (just like htmx, this lsp should be backend language agnostic).
-  The tag `hx-lsp` isn't an official htmx tag and I don't think we should be adding them (even though its "inert"). Also, this tag/data is sent down to the client and while I can't imagine that it would "leak" sensitive data, it's something that has stuck in the back of my head.

If there are incremental improvements this PR can be broken down into, that would be extremely helpful for reviewing. Curious to hear other peoples thoughts!

**Comment by uros-5** [comment]
> As @cristianoliveira said, please cherry pick out the fix into another PR. Happy to get that merged!
> 
> Honestly, this PR scares me `:)`
> 
>     * Its effectively an entire rewrite (has enough line changes to be) and is way to big to be able to effectively review (I suck at reviews, so YMMV). Anything over ~150 lines is where my eyes start to glaze over.
> 
>     * This feels like major feature creep. I really like the idea of having `gd` let me jump to the definition of an endpoint from the frontend, but I'm not sure that this lsp is the place to do so. If there are other solutions that provides tagging and jump points, let use that. Also, having to try and handle issue requests of "please add language x" just isn't manageable (just like htmx, this lsp should be backend language agnostic).
> 
>     * The tag `hx-lsp` isn't an official htmx tag and I don't think we should be adding them (even though its "inert"). Also, this tag/data is sent down to the client and while I can't imagine that it would "leak" sensitive data, it's something that has stuck in the back of my head.
> 
> 
> If there are incremental improvements this PR can be broken down into, that would be extremely helpful for reviewing. Curious to hear other peoples thoughts!

Yeah it's big rewrite. 

> handle issue requests of "please add language x"

I can manage that, query for tree sitter just requires `hx@` comment.

> Also, this tag/data is sent down to the client and while I can't imagine that it would "leak" sensitive data

For production just remove this attribute(there are probably tools for that) :)

**Comment by ThePrimeagen** [comment]
I am glad to review this, but I'm still away for a moment.

**Comment by ThePrimeagen** [comment]
Still away :)

Don't forget to resolve the conflict!

**Comment by itepastra** [comment]
I cannot talk about the code, but the updates to the documentation itself are essentially what I wanted to do when I started making them.

**Comment by ThePrimeagen** [comment]
I am on @MNThomson team about making this into a few PRs. It's really hard to follow when everything has been touched.

Second, could you please have your work finished before you're pushing few times a day

**Comment by uros-5** [comment]
> I am on @MNThomson team about making this into a few PRs. It's really hard to follow when everything has been touched.
> 
> Second, could you please have your work finished before you're pushing few times a day

This is complete rewrite. And there are many new things added. I understand why it's so hard to follow, that's why it's best to move this to another repo([comment](https://github.com/ThePrimeagen/htmx-lsp/pull/36#discussion_r1421707650)).. 

Binary will just get bigger after we add more TreeSitters(didn't think about that, sorry). Right now binary size of your crate(htmx-lsp) is 4mb with opt-level 3. Mine is 7mb.

It's big because of new added features like:
- goto definition
- references
- goto implementation
- code actions 
- did save
- TreeSitters(html, python, go, javascript, rust) + queries for them.
- Incremental parsing for TreeSitter
- backend/frontend tags

Second problem is VsCode support. I haven't tested there. It's just too much work for basic things. Just look rust-analyzer or tailwindcss and how ugly that is.
Rust analyzer has over 2k lines in [`package.json`](https://github.com/rust-lang/rust-analyzer/blob/master/editors/code/package.json). Tailwind(only two features, hover and completion) has 800 lines of [`extension.ts`](https://github.com/tailwindlabs/tailwindcss-intellisense/blob/master/packages/vscode-tailwindcss/src/extension.ts). 

Many things are probably confusing so I added documentation, maybe that could help little bit.

Before build delete old Cargo.lock and run:

```sh
cargo build
cargo doc --open
``` 

If it didn't help, simply close this PR.

Happy New Year!

**Comment by ThePrimeagen** [comment]
this is just too much.  do as everyone has stated about reducing the scope and we can chat.  i am intentionally moving slow here to prevent issues and this is large large change.
